File: src/bin/0143-reorder-list.cpp
// @leetup=info id=143 lang=cpp slug=reorder-list

#include <algorithm>
#include <iostream>
#include <map>
#include <set>
#include <tuple>
#include <vector>

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

// @leetup=code

// Ref: ./0143-reorder-list.py

class Solution {
 public:
  void reorderList(ListNode* head) {
    if (!head) {
      return;
    }

    // Find a middle point.
    ListNode* slow = head;
    ListNode* fast = head;
    while (fast && fast->next) {
      slow = slow->next;
      fast = fast->next->next;
    }

    // Reverses the 2nd half.
    ListNode* prev = nullptr;
    ListNode* curr = slow->next;
    while (curr) {
      ListNode* next_node = curr->next;
      curr->next = prev;
      prev = curr;
      curr = next_node;
    }

    // End the 1st half.
    slow->next = nullptr;

    // Merge the two lists: the 1st half and 2nd half.
    ListNode* a = head;
    ListNode* b = prev;
    while (b) {
      ListNode* next_node = a->next;
      a->next = b;
      a = b;
      b = next_node;
    }

  }
};
// @leetup=code

int main() {
  std::cout << "Hello, world!" << std::endl;
  return 0;
}

File: src/bin/0143-reorder-list.py
# @leetup=info id=143 lang=python3 slug=reorder-list


from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# @leetup=code
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """

        if not head:
            return

        # Step 1: find middle
        # [1,2,3,4,5]
        #  ^ slow
        #  ^ fast
        slow, fast = head, head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
        # [1,2,3,4,5]
        #      ^ slow
        #          ^ fast

        # Step 2: reverse second half
        prev, curr = None, slow.next
        # [1,2,3,4,5]
        #        ^ cur
        while curr:
            next_node = curr.next
            curr.next = prev
            prev, curr = curr, next_node
        # [1,2,3,4,5]
        #      ^ slow
        # [4 <- 5]
        #       ^ prev

        slow.next = None
        # [1,2,3]
        #      ^ slow

        # Step 3: merge lists
        head1, head2 = head, prev
        # [1,2,3]
        #  ^ head1
        # [5, 4]
        #  ^ head2
        while head2:
            next_node = head1.next
            # [1,2,3]
            #  ^ head1
            #    ^ next_node
            # [5, 4]
            #  ^ head2
            head1.next = head2
            # [1,5,4]
            #  ^ head1
            #    ^ head2
            #   [2, 3]
            #    ^ next_node
            head1 = head2
            head2 = next_node
            # [1,5,4]
            #    ^ head1
            # [2,3]
            #  ^ head2


# @leetup=code
